.github/workflows/container.reusable.yml

```
@@ -0,0 +1,91 @@
+name: Build container image
+on:
+  workflow_call:
+    inputs:
+      name:
+        description: Name of the image to build
+        required: true
+        type: string
+      tag:
+        description: Tag of the image to build
+        required: true
+        type: string
+      load:
+        description: Whether to load the image after building it
+        default: false
+        required: false
+        type: boolean
+      push:
+        description: Whether to push the image after building it
+        default: false
+        required: false
+        type: boolean
+      latest:
+        description: Whether to also build (and push, if push == true) the `latest` tag
+        default: false
+        required: false
+        type: boolean
+      artifact:
+        description: Name of the artifact where the binaries have been uploaded
+        required: false
+        default: nri-kubernetes
+        type: string
+      platforms:
+        description: Comma-separated platforms to build for
+        required: false
+        default: linux/amd64,linux/arm64,linux/arm
+        type: string
+
+jobs:
+  container:
+    name: Build container image
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Download all artifacts from build job
+        uses: actions/download-artifact@v2
+        with:
+          name: ${{ inputs.artifact }}
+          path: bin
+
+      # It is common for branch names to have / characters, which cannot be used as container tags.
+      - name: Sanitize tag
+        id: tag
+        run: |
+          tag="${{ inputs.tag }}"
+          echo "::set-output name=name::${tag//\//-}"
+
+      - name: Build container images
+        run: |
+          load=""
+          if [[ "${{ inputs.load}}" == "true" ]]; then
+            load="--load"
+          fi
+
+          docker buildx build $load --platform=${{ inputs.platforms }} \
+            -t ${{ inputs.name }}:${{ steps.tag.outputs.name }} \
+            .
+
+      - uses: docker/login-action@v1
+        if: ${{ inputs.push }}
+        with:
+          username: ${{ secrets.FSI_DOCKERHUB_USERNAME }}
+          password: ${{ secrets.FSI_DOCKERHUB_TOKEN }}
+
+      - name: Push versioned images
+        if: ${{ inputs.push }}
+        run: |
+          docker buildx build --push --platform="${{ inputs.platforms }}" \
+            -t "${{ inputs.name }}:${{ steps.tag.outputs.name }}" \
+            .
+      - name: Push :latest images
+        if: ${{ inputs.push && inputs.latest }}
+        run: |
+          docker buildx build --push --platform=${{ inputs.platforms }} \
+            -t ${{ inputs.name }}:latest \
+            .
```

```suggestion
      - name: Push versioned images
        if: ${{ inputs.push && ! inputs.latest }}
        run: |
          docker buildx build --push --platform="${{ inputs.platforms }}" \
            -t "${{ inputs.name }}:${{ steps.tag.outputs.name }}" \
            .

      - name: Push versioned images :latest images
        if: ${{ inputs.push && inputs.latest }}
        run: |
          docker buildx build --push --platform=${{ inputs.platforms }} \
            -t "${{ inputs.name }}:${{ steps.tag.outputs.name }}" \
            -t ${{ inputs.name }}:latest \
            .
```

It should be faster and does not test if layers are cached or not.

---

Is the only change here the `if` line? Github is not highlighting anything and my eyes are old.

If it is, `inputs.latest` is intended to mean "push `:latest` _too_", not "push `:latest` only".

---

The `if` line and that it is adding two `--target`/`-t` in the latest.

Having them separated into two commands can lead to cache miss and rebuilding containers twice, having a `latest` container virtually different to the one with a `vX.Y.Z` tag.